Finished implementation of new Role Permissions

XanderDJ opened the pull request:
As described in Issue #163 I've implemented a new RolePermissions system that expanded the previous one present in Guild.hs. Added as an internal type to Types and then exported forwards so taht it's available when importing Discord.Types. New representation of RolePermissions through a newtype that wraps around an Integer. The other functions are all bit operations mostly.

XanderDJ commented:
Also fixed random stuff I've encountered here and there like typos

L0neGamer commented:
I said I would review in the morning but I have some comments now.
You create a map from an enum to that enum and its bit representation, and then use the theoretically unsafe M.! to access it. Why not just create a function that does the same thing? This would remove the need for a map constant in code, as well as reduce the requirement of M.!.

I have more comments but they can wait for tomorrow.

XanderDJ commented:
> I said I would review in the morning but I have some comments now. You create a map from an enum to that enum and its bit representation, and then use the theoretically unsafe M.! to access it. Why not just create a function that does the same thing? This would remove the need for a map constant in code, as well as reduce the requirement of M.!.
> 
I had already changed this since I've refactored it a bit. I squashed the commits hence why it's only 1 commit

L0neGamer commented on src/Discord/Internal/Rest/Guild.hs:
```diff
@@ -186,7 +186,7 @@ data ModifyGuildRoleOpts = ModifyGuildRoleOpts
   toJSON ModifyGuildRoleOpts{..} = objectFromMaybes
                        ["name" .=? modifyGuildRoleOptsName,
-                        "permissions" .=? modifyGuildRoleOptsPermissions,
+                        "permissions" .=? fmap (T.pack . show) modifyGuildRoleOptsPermissions,
```
This indicates that we may want to create a ToJSON instance for RolePermissions

L0neGamer commented on src/Discord/Internal/Types/Guild.hs:
```diff
@@ -276,78 +274,16 @@ instance FromJSON Role where
          <*> o .: "hoist"
          <*> o .: "position"
-         <*> o .: "permissions"
+         <*> fmap (readMaybe . T.unpack) (o .: "permissions") .!= RolePermissions 0
```
the use of readmaybe here indicates that we may want a FromJSON instance for RolePermissions. Further, if we can't read the string they throw at us, we should fail, so the use of readMaybe here isn't advised.

L0neGamer commented on src/Discord/Internal/Types/RolePermissions.hs:
```diff
@@ -0,0 +1,119 @@
+    go [] = []
+    go (x : xs) = case roleIdToRole g x of
+      Nothing -> [False] <> go xs
+      Just a -> [p `hasRolePermission` rolePerms a] <> go xs
```
minor: newline at end of file please!

L0neGamer commented on src/Discord/Internal/Types/RolePermissions.hs:
```diff
@@ -0,0 +1,119 @@
+hasGuildMemberPermission :: Guild -> GuildMember -> PermissionFlag -> Bool
+hasGuildMemberPermission g gm p = or $ go (memberRoles gm)
+  where
+    go [] = []
+    go (x : xs) = case roleIdToRole g x of
+      Nothing -> [False] <> go xs
+      Just a -> [p `hasRolePermission` rolePerms a] <> go xs
```
Why not something like
```haskell
go [] = False
go (x:xs) = case roleIdToRole g of
    Nothing -> go xs
    Just a -> (p `hasRolePermission` rolePerms a) || go xs
```

Then we don't need to try and construct a list and then or through it.

Alternatively, the lists could be constructed by prepending the elements you're suggesting to the front.

XanderDJ replied:
This function was one of aquarials one, I didn't want to touch it tbh I just changed it so that it worksw with the new types and that's basically it. I'll implement the feedback later

L0neGamer commented on src/Discord/Internal/Types/RolePermissions.hs:
```diff
@@ -0,0 +1,119 @@
+combinePermissions = foldr ((.|.) . permissionBits) 0
+
+-- | Check if any Role of an GuildMember has the needed permission
+--   If the result of roleIdToRole is Nothing, it appends an "False"
```
> it appends an "False"

to

> it prepends a `False`

L0neGamer reviewed (changes requested):
I'm generally happy that this will work as intended, but I have a couple of comments I'd like addressed first.

XanderDJ commented:
Implemented all feedback
